lib/multi.c

```diff
@@ -2019,7 +2020,32 @@ static CURLMcode multi_runsingle(struct Curl_multi *multi,
       break;
 
     case MSTATE_DO:
-      if(data->set.connect_only) {
+      callback_failure = FALSE;
+      if(data->set.fprereq) {
+        int prereq_rc;
+
+        /* call the prerequest callback function */
+        Curl_set_in_callback(data, true);
+        prereq_rc = data->set.fprereq(data->set.prereq_userp,
+                                      data->info.conn_primary_ip,
+                                      data->info.conn_local_ip,
+                                      data->info.conn_primary_port,
+                                      data->info.conn_local_port);
+        Curl_set_in_callback(data, false);
+        if(prereq_rc != CURL_PREREQFUNC_OK) {
+          failf(data, "operation aborted by pre-request callback");
+          callback_failure = TRUE;
+        }
+      }
+
+      if(callback_failure) {
+        /* failure in pre-request callback - don't do any other processing  */
+        result = CURLE_ABORTED_BY_CALLBACK;
+        Curl_posttransfer(data);
+        multi_done(data, result, FALSE);
+        stream_error = TRUE;
+      }
+      else if(data->set.connect_only) {
```

Comment:
As you want to stop the execution if the pre-request callback fails we could break early and avoid modifying the control flow? (maybe a bit nitpicking :)) 
```c
case MSTATE_DO:
if(data->set.fprereq) {
...
  if (failure){
    break;
  }
}
if(data->set.connect_only) {
...
}
else {
...
}
```


```suggestion
      if(prereq_rc != CURL_PREREQFUNC_OK) {
          failf(data, "operation aborted by pre-request callback");
          /* failure in pre-request callback - don't do any other processing  */
          result = CURLE_ABORTED_BY_CALLBACK;
          Curl_posttransfer(data);
          multi_done(data, result, FALSE);
          stream_error = TRUE;
          break;
        }
      }
      
      if(data->set.connect_only) {
```

Reply:
Seems like a reasonable thing to do. I was trying hard not to mess up any of the existing call flow but yeah, an early break makes sense.